### PR TITLE
Metadata show filtered cell numbers

### PR DESCRIPTION
**Pull request by cosa65**
# Background
#### Link to issue 
https://biomage.atlassian.net/browse/BIOMAGE-1061

#### Link to staging deployment URL 
https://ui-martinfosco-ui421.scp-staging.biomage.net/

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [x] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx [hunk 1]
```diff
@@ -32,6 +33,19 @@ const { Text } = Typography;
 
 const { TabPane } = Tabs;
 
+const generateIsFilteredCellIndicator = (geneExpressions) => {
+  const someGeneExpressionObj = _.find(geneExpressions, () => true);
```
Comment (robioson) [inline src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx]: What exactly are you doing here? `geneExpressions` is an array and my guess is you want an arbitrary element, right? Why not simply use `[0]`?

File: src/__test__/components/data-exploration/cell-sets-tool/CellSetsTool.test.jsx [hunk 2]
```diff
@@ -334,4 +347,36 @@ describe('CellSetsTool', () => {
...
+  //   );
+  //   waitForComponentToPaint(component);
+  //   const tabs = component.find(Tabs);
+  //   const text = component.find('#selectedCellSets').first();
+  //   expect(text.text()).toEqual('3 cells selected');
+  //   tabs.props().onChange('metadataCategorical');
+  //   expect(text.text()).toEqual('4 cells selected');
+  //   tabs.props().onChange('cellSets');
+  //   expect(text.text()).toEqual('3 cells selected');
+  // });
```
Comment (robioson) [inline src/__test__/components/data-exploration/cell-sets-tool/CellSetsTool.test.jsx]: This should either be deleted or uncommented! (blocking)

**Review by robioson — DISMISSED** [review]
It's so good to get this fixed, only a couple of comments

File: src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx [hunk 3]
```diff
@@ -40,8 +50,29 @@ const CellSetsTool = (props) => {
...
+  const genes = useSelector(
+    (state) => state.genes,
+  );
+
+  const filteredCells = useRef(new Set());
+
   const [activeTab, setActiveTab] = useState('cellSets');
 
+  useEffect(() => {
+    if (filteredCells.current.size) return;
```
Comment (ivababukova) [inline src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx]: what happens if the user goes back to data processing and changes the filters? Or if they decide to filter out the gene which expression we use in here to get filtered cells from? 
Will filteredCells get updated? Will another gene get selected?

File: src/__test__/components/data-exploration/cell-sets-tool/CellSetsTool.test.jsx [hunk 4]
```diff
@@ -334,4 +347,36 @@ describe('CellSetsTool', () => {
...
 });
```
Comment (ivababukova) [inline src/__test__/components/data-exploration/cell-sets-tool/CellSetsTool.test.jsx:413]: We are also missing out unit tests to actually test out if this filtered cells functionality works.

File: src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx [hunk 5]
```diff
@@ -1,37 +1,47 @@
...
 import PropTypes from 'prop-types';
 
-import {
-  Skeleton, Space,
-  Tabs,
-  Typography, Empty, Button, Alert,
-} from 'antd';
+import { Alert, Button, Empty, Skeleton, Space, Tabs, Typography } from 'antd';
 
 import { BlockOutlined, MergeCellsOutlined, SplitCellsOutlined } from '@ant-design/icons';
```
Comment (ivababukova) [inline src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx:8]: can you add unit tests for this new filtered cells functionality?

**Comment by robioson** [comment]
I've added tests. Still trying to figure out the issues where selecting all cells by metadata results in a count one fewer than selecting all clusters. I noticed that the union of all clusters includes cell index 0, whereas the union of all samples does not. Still investigating this.

**Comment by ivababukova** [comment]
> I've added tests. Still trying to figure out the issues where selecting all cells by metadata results in a count one fewer than selecting all clusters. I noticed that the union of all clusters includes cell index 0, whereas the union of all samples does not. Still investigating this.

Code looks good. If you are confident that this bug is not caused by the changes that we are introducing with this PR, let's merge this PR and tackle the bug as part of a different ticket. What do you think, James?

**Review by ivababukova — APPROVED** [review]

File: src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx (hunk 3 above)
Comment (robioson) [reply src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx]: Very good catch, based on my testing you are correct that this was an issue. I removed the guard and confirmed that the `useEffect` only gets called once upon page load anyway.

**Comment by robioson** [comment]
I think this is the correct approach @ivababukova. I will create a ticket.

File: src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx (hunk 3 above)
Comment (ivababukova) [reply src/components/data-exploration/cell-sets-tool/CellSetsTool.jsx]: nice one!

**Pull request merged**